Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2024 Apple Inc.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "ProcessExtension.h"
+
+#if USE(EXTENSIONKIT)
+#import "AssertionCapability.h"
+
+#import <BrowserEngineKit/BECapability.h>
+#import <BrowserEngineKit/BENetworkingProcess.h>
+#import <BrowserEngineKit/BERenderingProcess.h>
+#import <BrowserEngineKit/BEWebContentProcess.h>
+
+namespace WebKit {
+
+ExtensionProcess::ExtensionProcess(BEWebContentProcess *process)
+    : m_process(process)
+{
+}
+
+ExtensionProcess::ExtensionProcess(BENetworkingProcess *process)
+    : m_process(process)
+{
+}
+
+ExtensionProcess::ExtensionProcess(BERenderingProcess *process)
+    : m_process(process)
+{
+}
+
+void ExtensionProcess::invalidate() const
+{
+    WTF::switchOn(m_process, [&] (std::monostate) {
+    }, [&] (RetainPtr<BEWebContentProcess> process) {
+        [process invalidate];
+    }, [&] (RetainPtr<BENetworkingProcess> process) {
+        [process invalidate];
+    }, [&] (RetainPtr<BERenderingProcess> process) {
+        [process invalidate];
+    });
```

Comment:
Shouldn't we set `m_process` to its monostate then?

Reply:
That is a very good point. I think this behavior is matching the previous behavior, where we would only invalidate the process object here, but deallocate the object when the ProcessLauncher is deleted.